napari/plugins/_npe2.py

```diff
@@ -55,7 +55,11 @@ def read(
             npe1_path, plugin_name=plugin
         )
     except ValueError as e:
-        if 'No readers returned data' not in str(e):
+        # Catch both messages raised by npe2
+        # https://github.com/napari/npe2/blob/main/src/npe2/io_utils.py#L168-L173
+        plugin_name_msg = f"Plugin {plugin!r} was selected to open"
+        no_plugin_name_msg = "No readers returned data"
```

Comment:
As I mentioned, I think we want to make sure we do raise the `Plugin was selected to open` error here - it's meant to indicate that the user selected a specific plugin, this plugin was tried, and it failed. Definitely we want to alert the user to this.

What I was missing before is that (because of the `npe1` compatibility code) this function **may** be passed a non-npe2 plugin, and in that case, we don't want to raise an error because we still have `npe1` plugins to try.

As such, I think we should instead add the following check to the top of this function (maybe below the `stack` assertion):

```python
# plugin was passed but is not an npe2 reader
if plugin and plugin not in get_readers():
    return None
```

I would also add a comment above the `"No readers returned data"` check:

```python
# plugin wasn't passed and no reader was found
if 'No readers returned data' not in str(e):
    raise
```

Reply:
Thanks for the explanation - that makes a lot more sense that what I thought was happening. I assumed that the suppressed errors would somehow be raised [here](https://github.com/napari/napari/blob/main/napari/_qt/dialogs/qt_reader_dialog.py#L293-L294)

I've added your suggestions and it works nicely!

Reply:
> I assumed that the suppressed errors would somehow be raised [here](https://github.com/napari/napari/blob/main/napari/_qt/dialogs/qt_reader_dialog.py#L293-L294)

Ahh yes, that's a GUI specific bit of code to allow users to select a different plugin if their saved plugin for a given path failed. The tests above occur without the GUI (or even a napari viewer), so that piece of code doesn't get a chance to handle anything. Even in the GUI scenario that error would've been swallowed, and I think we'd instead get whatever error (if any) was thrown at the end of trying `npe1` plugins.

Reply:
ah I see, thanks for the further info :)